plugin/trino-loki/src/main/java/io/trino/plugin/loki/LokiRecordSet.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.loki;
+
+import com.google.common.collect.ImmutableList;
+import io.github.jeschkies.loki.client.LokiClient;
+import io.github.jeschkies.loki.client.LokiClientException;
+import io.github.jeschkies.loki.client.model.Matrix;
+import io.github.jeschkies.loki.client.model.QueryResult;
+import io.github.jeschkies.loki.client.model.Streams;
+import io.trino.spi.TrinoException;
+import io.trino.spi.connector.RecordCursor;
+import io.trino.spi.connector.RecordSet;
+import io.trino.spi.type.Type;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public class LokiRecordSet
+        implements RecordSet
+{
+    private final List<LokiColumnHandle> columnHandles;
+    private final List<Type> columnTypes;
+
+    private final QueryResult result;
+
+    public LokiRecordSet(LokiClient lokiClient, LokiSplit split, List<LokiColumnHandle> columnHandles)
+    {
+        requireNonNull(lokiClient, "lokiClient is null");
+        requireNonNull(split, "split is null");
+
+        this.columnHandles = requireNonNull(columnHandles, "columnHandles is null");
+        ImmutableList.Builder<Type> types = ImmutableList.builder();
+        for (LokiColumnHandle column : columnHandles) {
+            types.add(column.type());
+        }
+        this.columnTypes = types.build();
+
+        Instant end = Instant.now();
+        Instant start = end.minus(Duration.ofHours(1));
+        if (split.end().getEpochSecond() != 0) {
+            end = split.end();
+        }
+        if (split.start().getEpochSecond() != 0) {
+            start = split.start();
+        }
+
+        final String query = split.query();
+        // Actually execute the query
+        // TODO: lazily parse
+        try {
+            this.result = lokiClient.rangeQuery(query, start, end);
+        }
+        catch (LokiClientException e) {
+            throw new TrinoException(LokiErrorCode.LOKI_UNKNOWN_ERROR, e);
+        }
+    }
+
+    @Override
+    public List<Type> getColumnTypes()
+    {
+        return columnTypes;
+    }
+
+    @Override
+    public RecordCursor cursor()
+    {
+        return switch (result.getData().getResult()) {
+            case Streams s -> new LokiStreamsRecordCursor(columnHandles, s);
+            case Matrix m -> new LokiMatrixRecordCursor(columnHandles, m);
```

Comment:
streams and matrix

Reply:
It's funny. I also prefer full names but in Go people seem to prefer short ones. I've changed it.